File: files/game/chessMaster/game/chineseChess/chessBoard/piece/horse.lua
next={}
function bound_check(i,j)
	if((i>=0 and j>=0)and(i<=8 and j<=9))then
		return true
	end	
	return false
end
function next_step(x,y,player,board)  
	next={}
	local i=x+1
	local j=y+2*player
	if bound_check(i,j)
	and(get_board(x,y+1*player)==0)
	and get_board(i,j)*player<=0 then
		table.insert(next,i)
		table.insert(next,j)
	end
	
	i=x+1
	j=y-2*player
	if bound_check(i,j)
	and(get_board(x,y-1*player)==0)
	and get_board(i,j)*player<=0 then
		table.insert(next,i)
		table.insert(next,j)
	end
	
	i=x-1
	j=y+2*player
	if bound_check(i,j)
	and(get_board(x,y+1*player)==0)
	and get_board(i,j)*player<=0 then
		table.insert(next,i)
		table.insert(next,j)
	end
	
	i=x-1
	j=y-2*player
	if bound_check(i,j)
	and(get_board(x,y-1*player)==0)
	and get_board(i,j)*player<=0 then
		table.insert(next,i)
		table.insert(next,j)
	end
--
	i=x+2
	j=y+1*player
	if bound_check(i,j)
	and(get_board(x+1,y)==0)
	and get_board(i,j)*player<=0 then
		table.insert(next,i)
		table.insert(next,j)
	end
	
	i=x+2
	j=y-1*player
	if bound_check(i,j)
	and(get_board(x+1,y)==0)
	and get_board(i,j)*player<=0 then
		table.insert(next,i)
		table.insert(next,j)
	end
	
	i=x-2
	j=y+1*player
	if bound_check(i,j)
	and(get_board(x-1,y)==0)
	and get_board(i,j)*player<=0 then
		table.insert(next,i)
		table.insert(next,j)
	end
	
	i=x-2
	j=y-1*player
	if bound_check(i,j)
	and(get_board(x-1,y)==0)
	and get_board(i,j)*player<=0 then
		table.insert(next,i)
		table.insert(next,j)
	end

	return next
end 

File: src/source/class/game/chessMaster/chessboard/chineseChess/piece/Horse.cpp
#include "class/game/chessMaster/chessboard/chineseChess/piece/Horse.h"
#include "class/display/draw/drawObject/AllDrawObjects.h"
#include "class/game/chessMaster/chessboard/ChessBoard.h"
namespace CM {

Horse::Horse() {
	draw_piece1=Display::AllDrawObjects::get_cur_object()->get("chineseChess/red_horse");
	draw_piece2=Display::AllDrawObjects::get_cur_object()->get("chineseChess/black_horse");
	weight=30;
}
Horse::~Horse() {

}
void Horse::next_step(CM::Board<short int> *chess_board,
		int x,int y,Tim::vector<CM::Step> &next_step,int player){
	CM::Step cur_step;
	cur_step.add_move(x,y,0,-1);
	cur_step.add_move(0,0,chess_board->get(x,y),1);
	int i,j;
	i=x+1;
	j=y+2;
	if(bound_check(i,j)&&chess_board->get(x,y+1)==0&&chess_board->get(i,j)*player<=0){
		cur_step.moves[1].x=i;
		cur_step.moves[1].y=j;
		next_step.push_back(cur_step);
	}
	i=x-1;
	j=y+2;
	if(bound_check(i,j)&&chess_board->get(x,y+1)==0&&chess_board->get(i,j)*player<=0){
		cur_step.moves[1].x=i;
		cur_step.moves[1].y=j;
		next_step.push_back(cur_step);
	}
	i=x+1;
	j=y-2;
	if(bound_check(i,j)&&chess_board->get(x,y-1)==0&&chess_board->get(i,j)*player<=0){
		cur_step.moves[1].x=i;
		cur_step.moves[1].y=j;
		next_step.push_back(cur_step);
	}
	i=x-1;
	j=y-2;
	if(bound_check(i,j)&&chess_board->get(x,y-1)==0&&chess_board->get(i,j)*player<=0){
		cur_step.moves[1].x=i;
		cur_step.moves[1].y=j;
		next_step.push_back(cur_step);
	}
	i=x+2;
	j=y+1;
	if(bound_check(i,j)&&chess_board->get(x+1,y)==0&&chess_board->get(i,j)*player<=0){
		cur_step.moves[1].x=i;
		cur_step.moves[1].y=j;
		next_step.push_back(cur_step);
	}
	i=x+2;
	j=y-1;
	if(bound_check(i,j)&&chess_board->get(x+1,y)==0&&chess_board->get(i,j)*player<=0){
		cur_step.moves[1].x=i;
		cur_step.moves[1].y=j;
		next_step.push_back(cur_step);
	}
	i=x-2;
	j=y+1;
	if(bound_check(i,j)&&chess_board->get(x-1,y)==0&&chess_board->get(i,j)*player<=0){
		cur_step.moves[1].x=i;
		cur_step.moves[1].y=j;
		next_step.push_back(cur_step);
	}
	i=x-2;
	j=y-1;
	if(bound_check(i,j)&&chess_board->get(x-1,y)==0&&chess_board->get(i,j)*player<=0){
		cur_step.moves[1].x=i;
		cur_step.moves[1].y=j;
		next_step.push_back(cur_step);
	}
}
} /* namespace CM */
